dix/req.c

```diff
@@ -1625,24 +1644,38 @@ static void dix_rop_completed(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 	struct m0_dix_rop_ctx *rop_del_phase2 = NULL;
 	bool                   del_phase2 = false;
 	struct m0_dix_cas_rop *cas_rop;
+	int64_t                min_success;
+	int64_t                successful_ops = 0;
 
 	(void)grp;
 	if (req->dr_type == DIX_NEXT)
 		m0_dix_next_result_prepare(req);
 	else {
+		min_success = req->dr_min_success;
+		M0_ASSERT(ergo(min_success < 1, min_success == -1));
+		if (min_success == -1) {
+			/*
+			 * The special value -1 indicates that we should use
+			 * (N+K)/2 + 1 for min_success.
+			 */
+			min_success = cas_rop_tlist_length(&rop->dg_cas_reqs)/2 + 1;
+		}
+		M0_ASSERT(min_success > 0);
+
+		successful_ops = m0_tl_reduce(cas_rop, scan, &rop->dg_cas_reqs, 0,
+				  + !!(scan->crp_creq.ccr_sm.sm_rc == 0));
+
 		/*
-		 * Consider DIX request to be successful if there is at least
-		 * one successful CAS request.
+		 * If enough operations succeeded to satisfy min_success,
+		 * ignore any operations that failed. If we don't have enough,
+		 * count all operations when deciding overall success (ensures the
+		 * overall operation is considered a failure).
 		 */
-		if (m0_tl_forall(cas_rop, cas_rop,
-				 &rop->dg_cas_reqs,
-				 cas_rop->crp_creq.ccr_sm.sm_rc != 0))
-			    dix_cas_rop_rc_update(cas_rop_tlist_tail(
-						  &rop->dg_cas_reqs), 0);
-
 		m0_tl_for (cas_rop, &rop->dg_cas_reqs, cas_rop) {
-			if (cas_rop->crp_creq.ccr_sm.sm_rc == 0)
+			if (successful_ops < min_success ||
+			    cas_rop->crp_creq.ccr_sm.sm_rc == 0) {
 				dix_cas_rop_rc_update(cas_rop, 0);
+			}
```

Comment:
Can you please elaborate more about the algorithms here?
What if the successful_ops is bigger than or equal to min_success ?
What if the successful_ops is smaller than min_success ?

Reply:
The rest of the patch looks fine.

Reply:
I added a longer description in a comment in the last commit

Reply:
With your updated comment, it is easier for me to understand the algorithm now.

According to your latest comment for the algorithm, can I write the code as the following?
```
                if (successful_ops >= min_success) {
                         /* quorum met. we only do fini() for all cas rop. */
                        m0_tl_for (cas_rop, &rop->dg_cas_reqs, cas_rop) {
                                m0_cas_req_fini(&cas_rop->crp_creq);
                        } m0_tl_endfor;
                } else {
                        /* quorum is not met. We call dix_cas_rop_rc_update for every cas op and then do fini() */
		        m0_tl_for (cas_rop, &rop->dg_cas_reqs, cas_rop) {
			        if (cas_rop->crp_creq.ccr_sm.sm_rc == 0) {
                                        /* If the cas rop failed to execute, its rc is already set */
				        dix_cas_rop_rc_update(cas_rop, 0);
			        }
                                m0_cas_req_fini(&cas_rop->crp_creq);
		        } m0_tl_endfor;
                }
```
I know the code seems a bit duplicated in this format, but it is  more readable, isn't it?

Again, the existing code is fine. I just want to have a better understanding of it.

Reply:
Not quite, it turns out `dix_cas_rop_rc_update()` does more than its name suggests (that's why I missed the leak the first time around). It also copies the value returned from the cas op into the dix request. We therefore always need to call it *unless* we're ignoring a result (to update the rc in the parent dix op in case of failure, or to copy the value from cas in  case of success). When ignoring results, we don't call it since there's no value to copy and we don't want to change the (zero) rc that will be passed back to the client.

It's also helpful to look at the commit that introduced the current behavior f2ba0d107310b2a795e6c3818b7fb4fa454eb10a

Reply:
I am confused, after rethinking a long time.

Let's ask @yeshpal-jain-seagate to review this patch with more detail.

Reply:
Let's take a close look at under what conditions the `dix_cas_rop_rc_update()` will be called.
When "successful_ops < min_success" or "cas_rop->crp_creq.ccr_sm.sm_rc == 0", the dix_cas_rop_rc_update() will be called.
Only when " successful_ops >= min_success" AND "cas_rop->crp_creq.ccr_sm.sm_rc != 0", the dix_cas_rop_rc_update() will be skipped.


------------------------------------------------------------------------------------------------------------------------------
                                                       | cas_rop->crp_creq.ccr_sm.sm_rc == 0 | cas_rop->crp_creq.ccr_sm.sm_rc != 0    |
-----------------------------------------------------------------------------------------------------------------------------|
successful_ops < min_success       |                    called                                 |                         called                              |
-----------------------------------------------------------------------------------------------------------------------------|
successful_ops >= min_success     |                   called                                  |                         NOT                               |
-----------------------------------------------------------------------------------------------------------------------------|


Is this right?

Reply:
Yes, exactly. That specific case (`successful_ops >= min_success` AND `cas_rop->crp_creq.ccr_sm.sm_rc != 0`) is a failed cas operation that we want to ignore since we've met the quorum requirement. Aside from that case, we should call `dix_cas_rop_rc_update()` to update the parent dix operation with the success/failure.

Reply:
Good. Now I understood the algorithm. Thank you, @trshaffer .